### PR TITLE
Storage/bugfix deserialization throws

### PR DESCRIPTION
**Pull request by anjbur**
* [ ] Has `CHANGELOG.md` been updated?
* [x] Are tests passing locally?
* [x] Are the files formatted correctly?
* [x] Did you add unit tests?
* [ ] Did you test your change with either the sample apps that are included in the repository or with a blank app that uses your change?
* [ ] Did you check UI tests on the sample app? They are not executed on CI.

## Description

Refactor document parsing so that deserialization errors don't result in an exception being thrown.

## Related PRs or issues

[AB#60021](https://msmobilecenter.visualstudio.com/web/wi.aspx?pcguid=27bf89de-9697-418a-accf-1aa125b22914&id=60021)

## Misc

Add what's missing, notes on what you tested, additional thoughts or questions.

### CODE REVIEW

File: sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/Storage.java [hunk 1]
```diff
@@ -805,6 +813,12 @@ public void completeFuture(Exception e) {
         mPendingCalls.remove(future);
     }
 
+    private synchronized <T> void completeFutureOnDocumentError(Document<T> doc, DefaultAppCenterFuture<Document<T>> future) {
+        AppCenterLog.error(Constants.LOG_TAG, "Failed to deserialize document: ", doc.getDocumentError().getError());
```
[the remark this answers is not in the document]
Comment (guperrot) [reply sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/Storage.java]: No need to use trailing space when logging an exception (it will add a newline)

File: sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/Utils.java [hunk 2]
```diff
@@ -34,7 +35,13 @@
     private static final JsonParser sParser = new JsonParser();
 
     static <T> Document<T> parseDocument(String cosmosDbPayload, Class<T> documentType) {
-        return parseDocument(sParser.parse(cosmosDbPayload).getAsJsonObject(), documentType);
+        JsonObject body;
+        try {
+            body = sParser.parse(cosmosDbPayload).getAsJsonObject();
+        } catch (Exception e) {
```
Comment (guperrot) [inline sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/Utils.java]: RuntimeException, it's not checked so we should not catch knives.

File: sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/Utils.java [hunk 3]
```diff
@@ -51,16 +58,16 @@ static String getEtag(String cosmosDbPayload) {
...
-        T document = sGson.fromJson(obj.get(Constants.DOCUMENT_FIELD_NAME), documentType);
         try {
+            T document = sGson.fromJson(obj.get(Constants.DOCUMENT_FIELD_NAME), documentType);
             return new Document<>(
                     document,
                     obj.get(Constants.PARTITION_KEY_FIELD_NAME).getAsString(),
                     obj.get(Constants.ID_FIELD_NAME).getAsString(),
                     obj.has(Constants.ETAG_FIELD_NAME) ? obj.get(Constants.ETAG_FIELD_NAME).getAsString() : "",
                     obj.get(Constants.TIMESTAMP_FIELD_NAME).getAsLong());
         } catch (Exception exception) {
```
Comment (guperrot) [inline sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/Utils.java]: why is the exception a checked one here? Isn't that RuntimeException?

File: sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/Utils.java [hunk 4]
```diff
@@ -70,8 +77,14 @@ static String getEtag(String cosmosDbPayload) {
     }
 
     public static <T> Page<T> parseDocuments(String cosmosDbPayload, Class<T> documentType) {
-        JsonObject objects = sParser.parse(cosmosDbPayload).getAsJsonObject();
-        JsonArray array = objects.get(Constants.DOCUMENTS_FILED_NAME).getAsJsonArray();
+        JsonArray array;
+        try {
+            JsonObject objects = sParser.parse(cosmosDbPayload).getAsJsonObject();
+            array = objects.get(Constants.DOCUMENTS_FILED_NAME).getAsJsonArray();
+        } catch (Exception e) {
```
Comment (guperrot) [inline sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/Utils.java]: RuntimException

Please read the details for example at https://www.geeksforgeeks.org/checked-vs-unchecked-exceptions-in-java/

File: sdk/appcenter-storage/src/test/java/com/microsoft/appcenter/storage/StorageTest.java [hunk 5]
```diff
@@ -372,6 +372,113 @@ public void replaceEndToEnd() throws JSONException {
         assertEquals(TEST_FIELD_VALUE, testDocument.test);
     }
 
+    @Test
+    public void readFailsToDeserializeDocumentDoesNotThrow() throws JSONException {
+        /* Mock http call to get token. */
```
Comment (guperrot) [inline sdk/appcenter-storage/src/test/java/com/microsoft/appcenter/storage/StorageTest.java]: add new lines before and after this comment

File: sdk/appcenter-storage/src/test/java/com/microsoft/appcenter/storage/StorageTest.java [hunk 6]
```diff
@@ -372,6 +372,113 @@ public void replaceEndToEnd() throws JSONException {
...
+        /* Get and verify document. Confirm the cache was not touched. */
+        assertNotNull(doc);
+        verifyNoMoreInteractions(mLocalDocumentStorage);
+        Document<String> testCosmosDocument = doc.get();
+        assertNotNull(testCosmosDocument);
+        assertTrue(testCosmosDocument.failed());
+    }
+
+    @Test
+    public void createFailsToDeserializeDocumentDoesNotThrow() throws JSONException {
```
Comment (guperrot) [inline sdk/appcenter-storage/src/test/java/com/microsoft/appcenter/storage/StorageTest.java]: Thats the JSONException from built in json library, not Gson, why is it here?

File: sdk/appcenter-storage/src/test/java/com/microsoft/appcenter/storage/StorageTest.java [hunk 7]
```diff
@@ -372,6 +372,113 @@ public void replaceEndToEnd() throws JSONException {
...
+        /* Make the call. Ensure deserialization error on Document by passing incorrect class type. */
+        PaginatedDocuments<String> docs = Storage.list(Constants.USER, String.class).get();
+
+        verifyNoMoreInteractions(mLocalDocumentStorage);
+
+        /* Verify the result is correct. */
+        assertNotNull(docs);
+        assertFalse(docs.hasNextPage());
+        Page<String> page = docs.getCurrentPage();
+        assertEquals(1, page.getItems().size());
```
Comment (guperrot) [inline sdk/appcenter-storage/src/test/java/com/microsoft/appcenter/storage/StorageTest.java]: assertNotNull(page.getItems()) before accessing it

Comment (guperrot) [reply sdk/appcenter-storage/src/test/java/com/microsoft/appcenter/storage/StorageTest.java]: comment that may apply to other tests

File: sdk/appcenter-storage/src/test/java/com/microsoft/appcenter/storage/StorageTest.java (hunk 6 above)
Comment (anjbur) [reply sdk/appcenter-storage/src/test/java/com/microsoft/appcenter/storage/StorageTest.java]: It's because`verifyTokenExchangeToCosmosDbFlow` mocks the http calls in a way that can throw (down on lines 278 and 307)

Comment (guperrot) [reply sdk/appcenter-storage/src/test/java/com/microsoft/appcenter/storage/StorageTest.java]: okay

File: sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/Storage.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (anjbur) [reply sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/Storage.java]: Got it, fixed.

File: sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/Utils.java (hunk 2 above)
Comment (anjbur) [reply sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/Utils.java]: Understood, change made.

File: sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/Utils.java (hunk 3 above)
Comment (anjbur) [reply sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/Utils.java]: Change made

File: sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/Utils.java (hunk 4 above)
Comment (anjbur) [reply sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/Utils.java]: Change made

File: sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/LocalDocumentStorage.java [hunk 8]
```diff
@@ -173,6 +173,11 @@ void resetDatabase() {
                 return new Document<>(new StorageException("Document was found in the cache, but it was expired. The cached document has been invalidated."));
             }
             Document<T> document = Utils.parseDocument(values.getAsString(DOCUMENT_COLUMN_NAME), documentType);
+            if (document.failed()) {
+                Throwable error = document.getDocumentError().getError();
+                AppCenterLog.error(LOG_TAG, "Failed to read from cache: ", error);
```
Comment (dhei) [inline sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/LocalDocumentStorage.java]: same as https://github.com/Microsoft/AppCenter-SDK-Android/pull/1075/files#r275086746

File: sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/Utils.java [hunk 9]
```diff
@@ -85,11 +100,11 @@ static String getEtag(String cosmosDbPayload) {
      * @param e Exception to display in the log.
      */
     public static synchronized void logApiCallFailure(Exception e) {
-        AppCenterLog.error(Constants.LOG_TAG, "Failed to call App Center APIs", e);
+        AppCenterLog.error(LOG_TAG, "Failed to call App Center APIs", e);
         if (!HttpUtils.isRecoverableError(e)) {
             if (e instanceof HttpException) {
                 HttpException httpException = (HttpException) e;
-                AppCenterLog.error(Constants.LOG_TAG, "Exception", httpException);
+                AppCenterLog.error(LOG_TAG, "Exception", httpException);
```
Comment (dhei) [inline sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/Utils.java]: I failed to see the reason why we only log unrecoverable `HttpException` in this util method. I don't see why we are filtering out recoverable `HTTPException`s.

Addressing this is out of scope of this PR, just want to bring it up.

Comment (guperrot) [reply sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/Utils.java]: Its duplicate log, http client already does this and I noticed I got logs twice in sasquatch. It's because we don't test e2e enough to notice.
